cmd/run.go

```diff
@@ -121,7 +122,12 @@ func NewRunCommand(stopCh <-chan struct{}) *cobra.Command {
 				return err
 			}
 
-			time.Sleep(time.Second * 3)
+			// try initializing provider
+			ctx := context.Background()
+			err = util.InitProvider(ctx, oidcOptions, stopCh)
```

Comment:
Can we try earlier before we start serving and init the whole of oidc-proxy but after offline validation?

Reply:
Yes, but I think I'll need to update the e2e tests--right now, they don't wait for oidc-proxy to be ready :slightly_smiling_face:

Reply:
Makes sense, think I have a dumb sleep in there atm...

Reply:
(Question: If the proxy starts serving only_after_ the offline validation, can the proxy itself be the target of the readiness probe?)

Reply:
So the readiness probe is independent of the serving - we create a listener on another port in a go routine `pkg/probe/probe.go`. If the probe in unreachable, its considered the same if the code is < 200 && >= 400.

Reply:
I used "readiness probe" to refer to the Kubernetes readiness probe, so I thought that kubelet could just check whether the proxy was serving :slightly_smiling_face:. 

But I see now that the probe can do more than just the proxy, e.g., it has both `/live` and `/ready` endpoints.

(I was also confused because the [demo manifest](https://github.com/jetstack/kube-oidc-proxy/blob/master/demo/yaml/kube-oidc-proxy.yaml) defines a TCP readiness check, so it doesn't use the `/ready` endpoint. I think it should be an HTTP readiness check. If that's right, I can fix that a separate PR)

Reply:
I see, no we expose a second port for exclusively the readiness probe, separate from the proxy serving itself.  
https://github.com/jetstack/kube-oidc-proxy/blob/182fa79a7854bd33f3827d89e222e10c57c4aed5/cmd/run.go#L52
https://github.com/jetstack/kube-oidc-proxy/blob/182fa79a7854bd33f3827d89e222e10c57c4aed5/pkg/probe/probe.go#L31

In theory, `/ready` is a valid endpoint that can be proxied to the backend so we can't re-use it. 

The health check is handled through the heptio library which seems to work quite well with exposing the correct `/ready` endpoint. We then just flip a bool value on the package struct to tell it to start serving 200s on requests manually 
https://github.com/jetstack/kube-oidc-proxy/blob/182fa79a7854bd33f3827d89e222e10c57c4aed5/cmd/run.go#L125

We don't currently have a liveness probe since there isn't much of a use case that I'm aware of..

Reply:
Now we try to init the provider before running the proxy. The e2e test verifies proxy is ready before using it.